File: meshlink-sample/android/app/src/main/kotlin/ch/trancee/meshlink/proof/android/ProofGattBenchmarkProtocol.kt
package ch.trancee.meshlink.proof.android

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID

internal object ProofGattBenchmarkProtocol {
    internal val SERVICE_UUID: UUID = UUID.fromString("4d455348-1001-1000-8000-000000000000")
    internal val WRITE_CHARACTERISTIC_UUID: UUID =
        UUID.fromString("4d455348-1002-1000-8000-000000000000")
    internal val ACK_CHARACTERISTIC_UUID: UUID =
        UUID.fromString("4d455348-1003-1000-8000-000000000000")
    internal val CCCD_UUID: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
    internal const val MANUFACTURER_ID: Int = 0xFFFF

    private const val ADVERTISEMENT_VERSION: Byte = 1
    private const val START_FRAME_TYPE: Byte = 0x01
    private const val DATA_FRAME_TYPE: Byte = 0x02
    private const val ACK_FRAME_TYPE: Byte = 0x11
    private const val TOKEN_BYTES: Int = 8
    private const val START_FRAME_BYTES: Int = 1 + TOKEN_BYTES + Int.SIZE_BYTES
    private const val ACK_FRAME_BYTES: Int = 1 + TOKEN_BYTES + Int.SIZE_BYTES
    private const val DATA_FRAME_HEADER_BYTES: Int = 1 + TOKEN_BYTES

    internal fun advertisementTag(appId: String): ByteArray {
        val appHash = foldedAppHash(appId)
        return byteArrayOf(
            ADVERTISEMENT_VERSION,
            (appHash.toInt() and 0xFF).toByte(),
            ((appHash.toInt() ushr 8) and 0xFF).toByte(),
        )
    }

    internal fun matchesAdvertisement(data: ByteArray?, appId: String): Boolean {
        return data?.contentEquals(advertisementTag(appId)) == true
    }

    internal fun encodeAck(tokenHex: String, totalBytes: Int): ByteArray {
        val tokenBytes = tokenHex.decodeHexToken() ?: error("Invalid benchmark token: $tokenHex")
        return ByteBuffer.allocate(ACK_FRAME_BYTES)
            .order(ByteOrder.BIG_ENDIAN)
            .put(ACK_FRAME_TYPE)
            .put(tokenBytes)
            .putInt(totalBytes)
            .array()
    }

    internal fun decodeWriteFrame(payload: ByteArray): WriteFrame? {
        if (payload.isEmpty()) {
            return null
        }
        return when (payload[0]) {
            START_FRAME_TYPE -> decodeStartFrame(payload)
            DATA_FRAME_TYPE -> decodeDataFrame(payload)
            else -> null
        }
    }

    internal fun decodeAckFrame(payload: ByteArray): AckFrame? {
        if (payload.size != ACK_FRAME_BYTES || payload[0] != ACK_FRAME_TYPE) {
            return null
        }
        val tokenBytes = payload.copyOfRange(1, 1 + TOKEN_BYTES)
        val totalBytes =
            ByteBuffer.wrap(payload, 1 + TOKEN_BYTES, Int.SIZE_BYTES).order(ByteOrder.BIG_ENDIAN).int
        return AckFrame(tokenHex = tokenBytes.toHexToken(), totalBytes = totalBytes)
    }

    private fun decodeStartFrame(payload: ByteArray): StartFrame? {
        if (payload.size != START_FRAME_BYTES) {
            return null
        }
        val tokenBytes = payload.copyOfRange(1, 1 + TOKEN_BYTES)
        val totalBytes =
            ByteBuffer.wrap(payload, 1 + TOKEN_BYTES, Int.SIZE_BYTES).order(ByteOrder.BIG_ENDIAN).int
        if (totalBytes <= 0) {
            return null
        }
        return StartFrame(tokenHex = tokenBytes.toHexToken(), totalBytes = totalBytes)
    }

    private fun decodeDataFrame(payload: ByteArray): DataFrame? {
        if (payload.size <= DATA_FRAME_HEADER_BYTES) {
            return null
        }
        val tokenBytes = payload.copyOfRange(1, 1 + TOKEN_BYTES)
        return DataFrame(
            tokenHex = tokenBytes.toHexToken(),
            chunk = payload.copyOfRange(DATA_FRAME_HEADER_BYTES, payload.size),
        )
    }

    private fun foldedAppHash(appId: String): UShort {
        val hash = fnv1a32(appId.encodeToByteArray())
        val folded = (((hash ushr 16) xor (hash and 0xFFFF)) and 0xFFFF).let { if (it == 0) 1 else it }
        return folded.toUShort()
    }

    private fun fnv1a32(bytes: ByteArray): Int {
        var hash = 0x811c9dc5.toInt()
        bytes.forEach { byte ->
            hash = hash xor (byte.toInt() and 0xFF)
            hash *= 0x01000193
        }
        return hash
    }

    private fun String.decodeHexToken(): ByteArray? {
        if (length != TOKEN_BYTES * 2) {
            return null
        }
        val bytes = ByteArray(TOKEN_BYTES)
        for (index in bytes.indices) {
            val value = substring(index * 2, index * 2 + 2).toIntOrNull(16) ?: return null
            bytes[index] = value.toByte()
        }
        return bytes
    }

    private fun ByteArray.toHexToken(): String {
        return joinToString(separator = "") { byte ->
            (byte.toInt() and 0xFF).toString(16).padStart(2, '0')
        }
    }

    internal sealed interface WriteFrame

    internal data class StartFrame(val tokenHex: String, val totalBytes: Int) : WriteFrame

    internal data class DataFrame(val tokenHex: String, val chunk: ByteArray) : WriteFrame

    internal data class AckFrame(val tokenHex: String, val totalBytes: Int)
}

File: meshlink-sample/ios/ProofApp/ProofGattBenchmarkProtocol.swift
import CoreBluetooth
import Foundation

enum ProofGattBenchmarkProtocol {
    static let serviceUUID = CBUUID(string: "4d455348-1001-1000-8000-000000000000")
    static let writeCharacteristicUUID = CBUUID(string: "4d455348-1002-1000-8000-000000000000")
    static let ackCharacteristicUUID = CBUUID(string: "4d455348-1003-1000-8000-000000000000")
    static let manufacturerPrefix = Data([0xff, 0xff])

    private static let advertisementVersion: UInt8 = 1
    private static let startFrameType: UInt8 = 0x01
    private static let dataFrameType: UInt8 = 0x02
    private static let ackFrameType: UInt8 = 0x11
    private static let tokenBytes = 8

    struct AckFrame {
        let tokenHex: String
        let totalBytes: Int
    }

    static func advertisementTag(appId: String) -> Data {
        let appHash = foldedAppHash(appId)
        return Data([
            advertisementVersion,
            UInt8(appHash & 0x00ff),
            UInt8((appHash & 0xff00) >> 8),
        ])
    }

    static func matchesAdvertisement(_ data: Data?, appId: String) -> Bool {
        var expected = manufacturerPrefix
        expected.append(advertisementTag(appId: appId))
        return data == expected
    }

    static func encodeStart(tokenHex: String, totalBytes: Int) -> Data {
        var data = Data([startFrameType])
        data.append(tokenData(tokenHex))
        appendUInt32(UInt32(totalBytes), to: &data)
        return data
    }

    static func encodeData(tokenHex: String, chunk: Data) -> Data {
        var data = Data([dataFrameType])
        data.append(tokenData(tokenHex))
        data.append(chunk)
        return data
    }

    static func decodeAck(_ data: Data) -> AckFrame? {
        guard data.count == 1 + tokenBytes + MemoryLayout<UInt32>.size, data.first == ackFrameType else {
            return nil
        }
        let tokenHex = hexToken(from: data.subdata(in: 1..<(1 + tokenBytes)))
        let totalBytes = data[(1 + tokenBytes)...].reduce(0) { partial, next in
            (partial << 8) | Int(next)
        }
        return AckFrame(tokenHex: tokenHex, totalBytes: totalBytes)
    }

    static func chunkPayloadBytes(maxWriteValueLength: Int) -> Int {
        max(1, maxWriteValueLength - (1 + tokenBytes))
    }

    private static func tokenData(_ tokenHex: String) -> Data {
        precondition(tokenHex.count == tokenBytes * 2, "Expected a 16-character benchmark token")
        return Data(stride(from: 0, to: tokenHex.count, by: 2).map { index in
            let start = tokenHex.index(tokenHex.startIndex, offsetBy: index)
            let end = tokenHex.index(start, offsetBy: 2)
            return UInt8(tokenHex[start..<end], radix: 16) ?? 0
        })
    }

    private static func hexToken(from data: Data) -> String {
        data.map { byte in String(format: "%02x", byte) }.joined()
    }

    private static func appendUInt32(_ value: UInt32, to data: inout Data) {
        data.append(UInt8((value >> 24) & 0xff))
        data.append(UInt8((value >> 16) & 0xff))
        data.append(UInt8((value >> 8) & 0xff))
        data.append(UInt8(value & 0xff))
    }

    private static func foldedAppHash(_ appId: String) -> UInt16 {
        let hash = fnv1a32(Array(appId.utf8))
        let folded = ((hash >> 16) ^ (hash & 0xffff)) & 0xffff
        return UInt16(folded == 0 ? 1 : folded)
    }

    private static func fnv1a32(_ bytes: [UInt8]) -> UInt32 {
        var hash: UInt32 = 0x811c9dc5
        for byte in bytes {
            hash ^= UInt32(byte)
            hash &*= 0x01000193
        }
        return hash
    }
}
